### PR TITLE
build: New option 'disable_hdrmd5_check' to ignore hdrmd5 mismatches

### PR DESCRIPTION
**Pull request by dmach**
Fixes: #1230

### CODE REVIEW

**Review by Firstyear — APPROVED** [review]
Looks good to me :)

File: osc/fetch.py [hunk 1]
```diff
@@ -221,12 +221,22 @@ def run(self, buildinfo):
                 cached += 1
                 if not i.name.startswith('container:') and i.pacsuffix != 'rpm':
                     continue
+
+                hdrmd5_is_valid = True
                 if i.hdrmd5:
                     if i.name.startswith('container:'):
                         hdrmd5 = dgst(i.fullfilename)
+                        if not hdrmd5 or hdrmd5 != i.hdrmd5:
+                            hdrmd5_is_valid = False
```
Comment (adrianschroeter) [inline osc/fetch.py:230]: pitnick: could be simplified to

if hdrmd5 != i.hdrmd5:

because None is also not equal, no?

File: osc/fetch.py [hunk 2]
```diff
@@ -221,12 +221,22 @@ def run(self, buildinfo):
...
                     else:
                         hdrmd5 = packagequery.PackageQuery.queryhdrmd5(i.fullfilename)
-                    if not hdrmd5 or hdrmd5 != i.hdrmd5:
+                        if not hdrmd5 or hdrmd5 != i.hdrmd5:
+                            # FIXME: We're currently ignoring the hdrmd5 verification of rpm packages.
+                            # look for "HDRMD5 MISMATCH EXPLANATION" below
+                            print("Warning: Ignoring a hdrmd5 mismatch for %s: %s (actual) != %s (expected)" % (i.fullfilename, hdrmd5, i.hdrmd5))
+                            hdrmd5_is_valid = True
```
Comment (adrianschroeter) [inline osc/fetch.py:237]: this should be really only done via a switch. Or one looks endless for issues which are not reproducible locally due to broken noarch rpms (which are in fact not noarch)

File: osc/fetch.py [hunk 3]
```diff
@@ -258,19 +268,26 @@ def run(self, buildinfo):
...
+                        # OBS builds the noarch packages once per binary arch.
+                        # Such noarch packages are supposed to be nearly identical across all build arches,
+                        # any discrepancy in the payload and dependencies is considered a packaging bug.
+                        # But to guarantee that the local builds work identically to builds in OBS,
+                        # using the arch-specific copy of the noarch package is required.
+                        # Unfortunatelly only one of the noarch packages gets distributed
+                        # and can be downloaded from a local mirror.
+                        # All other noarch packages are available through the OBS API only.
+                        # Since there is currently no information about hdrmd5 checksums of published noarch packages,
+                        # we download them, verify hdrmd5 and re-download the package from OBS API on mismatch.
```
Comment (adrianschroeter) [inline osc/fetch.py:281]: noarch are the most prominent examples, but the same can also happen for architecture depend packages. eg. when you have multiple candidates in a project or when someone is messing around with the source history or the release number handling in a way that it is not increasing.

File: osc/fetch.py [hunk 4]
```diff
@@ -258,19 +268,26 @@ def run(self, buildinfo):
...
                         #
-                        # wbrown 2022 - is there a reason to keep these md5's at all? md5 is
-                        # broken from a security POV so these aren't a trusted source for validation
-                        # of the file content. They are often incorrect forcing download via the API
-                        # which for anyone outside the EU is excruciating. And when they are ignored
-                        # builds work and progress anyway? So what do they even do? What are they
-                        # for? They should just be removed.
+                        # FIXME: We're currently ignoring the hdrmd5 verification of rpm packages.
+                        # A proper fix requires OBS change such as described in
+                        # https://github.com/openSUSE/open-build-service/issues/13821
```
Comment (adrianschroeter) [inline osc/fetch.py:285]: please do not ignore it by default. The default should be really safe and reproduce the server side state as highest goal.

An optional switch to disable it should be implemented within a minute ....

**Review by adrianschroeter — CHANGES_REQUESTED** [review]
Please do not switch off the check by default, the reproducability of server side builds is really our highest goal with osc build.

File: osc/fetch.py (hunk 2 above)
Comment (dmach) [reply osc/fetch.py:237]: Agreed. `disable_hdrmd5_check` is going to be part of the changes I'm about to force-push.

File: osc/fetch.py (hunk 4 above)
Comment (dmach) [reply osc/fetch.py:285]: Yes, the `disable_hdrmd5_check` option will default to `0` (the original behavior when the exact checksums were required)

File: osc/fetch.py (hunk 3 above)
Comment (dmach) [reply osc/fetch.py:281]: Thanks for the details, I've used part of your text in the comment explaining the situation.

File: osc/fetch.py (hunk 1 above)
Comment (dmach) [reply osc/fetch.py:230]: Changed.

**Review by adrianschroeter — APPROVED** [review]

**Comment by bnavigator** [comment]
The description of this PR is incomplete. It does not only add an option to ignore the hdrmd5 but actually makes sure to download the correct package from API instead of published repos if ignoring is not an option.

**Pull request merged**